### PR TITLE
Add API tests against bleeding edge WC Rest API

### PR DESCRIPTION
**Pull request by bsessions85**
This PR adds in 27 new API tests that run against a site that is using the latest development version of the `woocommerce-rest-api`.

The tests will run daily by scheduled job that is configured in CircleCI config. The use Rest Assured to handle the calls and validation.

I think I hit every api call that is used in this project that hits the wc endpoints, but if you see any that I missed, please let me know.

### CODE REVIEW

**Comment by rachelmcr** [comment]
What do you think of adding something like a Slack notification, at least when tests fail? That could help with visibility of the results from these daily runs.

**Comment by anitaa1990** [comment]
@bsessions85, I noticed a few endpoints missing from the tests so thought I would add them here:

- [Update Product](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/a9561376c1093cd36113278fecbd2307cbe79312/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt#L358-L357) 
- [GET all refunds for an order](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/72b2b781475c830c24ba8044094472f5332913ea/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/refunds/RefundRestClient.kt#L112)
- [GET a single refund by ID](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/72b2b781475c830c24ba8044094472f5332913ea/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/refunds/RefundRestClient.kt#L88)
- [Refund an order](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/72b2b781475c830c24ba8044094472f5332913ea/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/refunds/RefundRestClient.kt#L65)
- [GET all payment gateways supported for the site.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/73fb317c2083c6d3afcb90e883e04a3819b1df4e/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/gateways/GatewayRestClient.kt#L52)
- [GET a single payment gateway by ID](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/73fb317c2083c6d3afcb90e883e04a3819b1df4e/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/gateways/GatewayRestClient.kt#L29)

Stats endpoints were missing as well but we use the V4 endpoints and not sure if that comes under the scope of this PR.

**Comment by bsessions85** [comment]
> What do you think of adding something like a Slack notification, at least when tests fail? That could help with visibility of the results from these daily runs.

@rachelmcr  Good idea. Do you have a place set up already where it would make sense to send the notification?

**Comment by rachelmcr** [comment]
> Do you have a place set up already where it would make sense to send the notification?

Since these tests are specific to WooCommerce it might make sense to send them to the `#mobile-woo` channel. @anitaa1990 would that make sense to you or is there a better channel (or should we create a new channel) for that kind of notification?

**Comment by bsessions85** [comment]
@anitaa1990 I have added those missing tests. Thanks for letting me know they were missing!

**Comment by anitaa1990** [comment]
>  @anitaa1990 would that make sense to you or is there a better channel (or should we create a new channel) for that kind of notification?

💯agree that #mobile-woo would be great in this case 👍

**Comment by rachelmcr** [comment]
Thanks, Anitaa! I set up a webhook saved in the CircleCI env var `SLACK_WOO_WEBHOOK` that we can use for that notification.

**Comment by bsessions85** [comment]
I have added the slack notification for failures. Thanks for setting up the var @rachelmcr !

**Comment by rachelmcr** [comment]
@bsessions85 Sorry for not getting back to this one sooner! I just realized it's still open and added myself as a reviewer so I won't lose track of it again. :)

Quick question: Is there a straightforward way to kick off a local test run for these API tests? I was thinking I should re-run them now to make sure they're still passing, but it seems like it'd also be useful for debugging failing tests in the future.

**Comment by bsessions85** [comment]
@rachelmcr To run the tests locally, you need to set the `API_TEST_OAUTH_KEY` and `API_TEST_SITE_URL` env variables and then run the tests using `./gradlew :tests:api:test`

Values for the env variables can be found in ss `Mobile API Test key`

File: tests/api/src/test/java/APITesting_WCOrder.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+            body("data", hasSize(1),
+                "data[0].tracking_number", equalTo("324567890234567")
+            );
+    }
+
+    @Test
+    public void canGetOrderShipmentProviders() {
+        given().
+            spec(this.mRequestSpec).
+            queryParam("path", "/wc/v4/orders/635/shipment-trackings/providers").
```
Comment (rachelmcr) [inline tests/api/src/test/java/APITesting_WCOrder.java:258]: As with the `orders/<id>/shipment-trackings` endpoint, it looks like `orders/<id>/shipment-trackings/providers` also isn't supported in `v4`. It would be good to double check the expected behavior but for now using `v3` resolves the failure I'm seeing:

```suggestion
            queryParam("path", "/wc/v3/orders/635/shipment-trackings/providers").
```

File: tests/api/src/test/java/APITesting_WCProduct.java [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        given().
+            spec(this.mRequestSpec).
+            header("Content-Type", ContentType.JSON).
+            queryParam("path", path).
+            queryParam("_method", method).
+            body(jsonObj.toString()).
+        when().
+            post().
+        then().
+            statusCode(200);
```
Comment (rachelmcr) [inline tests/api/src/test/java/APITesting_WCProduct.java:174]: This test is failing at this point for me, with the following error:

`Expected status code <200> but was <400>.`

Nothing is jumping out at me immediately as an issue here; do you see the same error on your end?

File: tests/api/src/test/java/APITesting_WCRefund.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        given().
+            spec(this.mRequestSpec).
+            header("Content-Type", ContentType.JSON).
+            queryParam("path", path).
+            queryParam("_method", method).
+            body(jsonObj.toString()).
+        when().
+            post().
+        then().
+            statusCode(200);
```
Comment (rachelmcr) [inline tests/api/src/test/java/APITesting_WCRefund.java:118]: This test is failing at this point for me, with this error:

`Expected status code <200> but was <500>.`

Have you seen that failure? I'm not sure what's going on there.

File: tests/api/src/test/java/APITesting_WCOrder.java [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        when().
+            post().
+        then().
+            statusCode(200).
+        extract().
+            path("data.tracking_id");
+        
+        // Delete tracking number    
+        method = "delete";
+        path = "/wc/v4/orders/634/shipment-trackings/" + tracking + "&_method=" + method;
```
[the remark this answers is not in the document]
Comment (bsessions85) [reply tests/api/src/test/java/APITesting_WCOrder.java]: yes, good idea

**Comment by bsessions85** [comment]
Thanks @rachelmcr . I knew most of these were failing on v4, because they aren't in the schema, so I switched them to v3 now. I also updated the path variable as suggested and the other failures you were seeing were I think issues with needing two paths in the same test to set up data. Right now they are all passing for me.
